### PR TITLE
Support saving Bfloat16 tensors for XLA/HPU

### LINKED ISSUE #77533: Support serialization of BFloat16 tensors for HPU
### 🐛 Describe the bug

PyTorch doesn't allow serializing HPU (and a few other backends which include XLA) tensors of type BFloat16.

Example with XLA (similar issue is observed with HPU tensors):

import torch
import torch_xla
import torch_xla.core.xla_model as xm

device = xm.xla_device()

t0 = torch.randn(2, 2, device=device).to(torch.bfloat16)
t1 = torch.randn(2, 2, device=device).to(torch.bfloat16)

tensors = (t0, t1)

torch.save(tensors, 'tensors.pt')

This example fails with -
TypeError                                 Traceback (most recent call last)
[<ipython-input-3-c58e3e38ac90>](https://localhost:8080/#) in <module>()
     10 tensors = (t0, t1)
     11 
---> 12 torch.save(tensors, 'tensors.pt')

3 frames
[/usr/local/lib/python3.7/dist-packages/torch/_tensor.py](https://localhost:8080/#) in _reduce_ex_internal(self, proto)
    201         #    and serialize them one by one.
    202         if self.device.type in ['xla', 'ort', 'mlc']:
--> 203             return (torch._utils._rebuild_device_tensor_from_numpy, (self.cpu().numpy(),
    204                                                                      self.dtype,
    205                                                                      str(self.device),

TypeError: Got unsupported ScalarType BFloat16

### Versions

PyTorch version: 1.11.0+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.22.4
Libc version: glibc-2.26

Python version: 3.7.13 (default, Apr 24 2022, 01:04:09)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.188+-x86_64-with-Ubuntu-18.04-bionic
Is CUDA available: False
CUDA runtime version: 11.1.105
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.21.6
[pip3] torch==1.11.0+cu113
[pip3] torch-xla==1.11
[pip3] torchaudio==0.11.0+cu113
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.12.0
[pip3] torchvision==0.12.0+cu113
[conda] Could not collect

### PR DESCRIPTION
**Pull request by sujoysaraswati**
Fixes #77533

### CODE REVIEW

**Comment by Lezcano** [comment]
Temptatively tagging @albanD as a reviewerd, but I don't know who would be a good reviewer for this. @albanD, feel free to untag yourself and tag a better reviewer.

File: torch/_tensor.py [hunk 1]
```diff
@@ -223,7 +223,11 @@ def _reduce_ex_internal(self, proto):
         #    `tolist()` converts every single element in the tensor into python objects
         #    and serialize them one by one.
         if self.device.type in ['xla', 'ort', 'mps', 'hpu']:
-            return (torch._utils._rebuild_device_tensor_from_numpy, (self.cpu().numpy(),
+            # Convert BFloat16 tesors to Float32 before conversion to numpy, as numpy doesn't
```
Comment (albanD) [inline torch/_tensor.py:226]: Can we add a test to make sure this works as expected?

**Review by albanD — COMMENTED** [review]
Thanks @Lezcano I can review this.

File: torch/_tensor.py (hunk 1 above)
Comment (sujoysaraswati) [reply torch/_tensor.py:226]: Hi @albanD , I have a test case in test_serialization.py but was not sure which device to use it for. CPU/CUDA serialization goes through a different path. I can try with xla, but I have a question. Can I add xla as a device to the test_serialization.py as is or do I need to co-ordinate with torch_xla?

Comment (albanD) [reply torch/_tensor.py:226]: cc @JackCaoG is that something that we could test on the xla CI to make sure we don't regress?

Comment (JackCaoG) [reply torch/_tensor.py:226]: Thanks @albanD, I can add a test on xla end.

Comment (sujoysaraswati) [reply torch/_tensor.py:226]: Thanks @JackCaoG

Comment (sujoysaraswati) [reply torch/_tensor.py:226]: @albanD - let me know if there is anything else required or if we can proceed with merging this patch.

**Review by albanD — APPROVED** [review]
Sounds good to me.
Any concerns about the increased memory footprint that will be double what it needs to be for bfloat16?

**Comment by sujoysaraswati** [comment]
> Sounds good to me. Any concerns about the increased memory footprint that will be double what it needs to be for bfloat16?

Thanks @albanD. As far as memory footprint is concerned, the conversion to FP32 is only during serialization and memory footprint increase is only limited during this part. Once the tensors are loaded back, they are converted back to BF16.

**Pull request closed without merging**